hoomd/hpmc/ExternalFieldHarmonic.h

```diff
@@ -0,0 +1,468 @@
+            for (size_t i = 0; i < N_particles; i++)
+                {
+                const size_t array_index = i * 4;
+                this->m_reference_orientations[i]
+                    = quat<Scalar>(rawdata[array_index],
+                                   vec3<Scalar>(rawdata[array_index + 1],
+                                                rawdata[array_index + 2],
+                                                rawdata[array_index + 3]));
+                }
+            }
+
+#ifdef ENABLE_MPI
+        if (this->m_sysdef->isDomainDecomposed())
+            {
+            bcast(m_reference_orientations, 0, m_exec_conf->getMPICommunicator());
+            }
+#endif
+        } // end setReferenceOrientations
+
+    //! Set symmetrically equivalent orientations from a (N_symmetry, 4) numpy array
+    void
+    setSymmetricallyEquivalentOrientations(const pybind11::array_t<double> equivalent_quaternions)
+        {
+        if (equivalent_quaternions.ndim() != 2)
+            {
+            throw std::runtime_error("The array must be of shape (N_sym, 4).");
+            }
+
+        const size_t N_sym = equivalent_quaternions.shape(0); // Number of equivalent orientations
+        const size_t dim = equivalent_quaternions.shape(1);
+        if (dim != 4)
+            {
+            throw std::runtime_error("The array must be of shape (N_sym, 4).");
+            }
+        const double* rawdata = static_cast<const double*>(equivalent_quaternions.data());
+        m_symmetry.resize(N_sym);
+        for (size_t i = 0; i < N_sym; i++)
+            {
+            const size_t array_index = i * 4;
+            this->m_symmetry[i] = quat<Scalar>(rawdata[array_index],
+                                               vec3<Scalar>(rawdata[array_index + 1],
+                                                            rawdata[array_index + 2],
+                                                            rawdata[array_index + 3]));
+            }
+
+#ifdef ENABLE_MPI
+        if (this->m_sysdef->isDomainDecomposed())
+            {
+            bcast(m_symmetry, 0, m_exec_conf->getMPICommunicator());
+            }
+#endif
+        } // end setSymmetricallyEquivalentOrientations
+
+    //! Get reference positions as a (N_particles, 3) numpy array
+    pybind11::array_t<Scalar> getReferencePositions() const
+        {
+        std::vector<size_t> dims(2);
+        dims[0] = this->m_reference_positions.size();
+        dims[1] = 3;
+        // the cast from vec3<Scalar>* to Scalar* is safe since vec3 is tightly packed without any
+        // padding. This also makes a copy so, modifications of this array do not effect the
+        // original reference positions.
+        const auto reference_array = pybind11::array_t<Scalar>(
+            dims,
+            reinterpret_cast<const Scalar*>(this->m_reference_positions.data()));
+        // This is necessary to expose the array in a read only fashion through C++
```

Comment:
If this returns a copy, why does it make it not writeable?
@b-butler does this need any special handling to interoperate with #1178?

Reply:
Deferring to @b-butler

Reply:
The reason this is not writable (I believe this is a copy paste from another updater) is because it would be deceiving as to what is happening. It is the copy-modify-write problem again. Given the fact that NumPy is not part of the Python standard library I did not create an array class that continually updates. It is possible though.

Reply:
That makes sense. This is attempting to give some notion of `const` or immutability with the read/modify/write mechanism.